bench/bench.ml

```diff
@@ -0,0 +1,480 @@
+    read_amplification_size : float;
+    write_amplification_calls : float;
+    write_amplification_size : float;
+  }
+  [@@deriving yojson]
+
+  let run ~with_metrics ~nb_entries f =
+    let _, time, stats = with_stats (fun () -> f ~with_metrics ()) in
+    let nb_entriesf = Float.of_int nb_entries in
+    let entry_sizef = Float.of_int entry_size in
+    let read_amplification_size =
+      Float.of_int stats.bytes_read /. (entry_sizef *. nb_entriesf)
+    in
+    let read_amplification_calls = Float.of_int stats.nb_reads /. nb_entriesf in
+    let write_amplification_size =
+      Float.of_int stats.bytes_written /. (entry_sizef *. nb_entriesf)
+    in
+    let write_amplification_calls =
+      Float.of_int stats.nb_writes /. nb_entriesf
+    in
+    let ops_per_sec = nb_entriesf /. time in
+    let mbs_per_sec = entry_sizef *. nb_entriesf /. 1_048_576. /. time in
+    {
+      time;
+      ops_per_sec;
+      mbs_per_sec;
+      read_amplification_calls;
+      read_amplification_size;
+      write_amplification_calls;
+      write_amplification_size;
+    }
+
+  let pp_result fmt result =
+    Format.fprintf fmt
+      "Total time: %f@\n\
+       Operations per second: %f@\n\
+       Mbytes per second: %f@\n\
+       Read amplification in syscalls: %f@\n\
+       Read amplification in bytes: %f@\n\
+       Write amplification in syscalls: %f@\n\
+       Write amplification in bytes: %f" result.time result.ops_per_sec
+      result.mbs_per_sec result.read_amplification_calls
+      result.read_amplification_size result.write_amplification_calls
+      result.write_amplification_size
+end
+
+let make_bindings_pool nb_entries =
+  Array.init nb_entries (fun _ ->
+      let k = Context.Key.v () in
+      let v = Context.Value.v () in
+      (k, v))
+
+let bindings_pool = ref [||]
+
+let absent_bindings = ref [||]
+
+module Index = struct
+  module Index = Index_unix.Private.Make (Context.Key) (Context.Value)
+
+  type index_v = fresh:bool -> readonly:bool -> string -> Index.t
+
+  type benchmark = with_metrics:bool -> index_v -> unit -> unit
+
+  let add_metrics =
+    let no_tags x = x in
+    fun () -> Metrics.add src no_tags (fun m -> m (Stats.get ()))
+
+  let write ~with_metrics ?(with_flush = false) bindings rw =
+    Array.iter
+      (fun (k, v) ->
+        Index.replace rw k v;
+        if with_flush then Index.flush rw;
+        if with_metrics then add_metrics ())
+      bindings
+
+  let read ~with_metrics bindings r =
+    Array.iter
+      (fun (k, (_ : Context.Value.t)) ->
+        ignore (Index.find r k : Context.Value.t);
+        if with_metrics then add_metrics ())
+      bindings
+
+  let read_absent ~with_metrics bindings r =
+    Array.iter
+      (fun (k, _) ->
+        try ignore (Index.find r k : Context.Value.t)
+        with Not_found -> if with_metrics then add_metrics ())
+      bindings
+
+  let write_random ~with_metrics v () =
+    v ~fresh:true ~readonly:false "" |> write ~with_metrics !bindings_pool
```

Comment:
Does it mean that the benchmarks also measures the time to create/open the index? I think we could add a setup function for each benchmark, to run before? What do you think?

Reply:
They do indeed, I thought it was part of what we wanted to measure (btw we have no dedicated bench for `v`). Would you prefer to have the `fresh` and `readonly` parameters inside the record and the function called before the timer starts to run?

Reply:
They do indeed, I thought it was part of what we wanted to measure (btw we have no dedicated bench for v). Would you prefer to have the fresh and readonly parameters inside the record and the function called before the timer starts to run?

Reply:
We don't have a benchmark for `v` and we could, indeed! But I prefer that the benchmarks for write and read only measure the performance of write and read respectively and not the time the db sets up for those.

Reply:
Gave it a try in https://github.com/mirage/index/pull/151/commits/3fda3e5c6e0312478ad2e33003ccb23cac0d6323